app/models/specific_tube_rack_creation.rb

```diff
@@ -0,0 +1,436 @@
+
+    create_racked_tubes(rack_attributes[:racked_tubes], new_tube_rack)
+
+    new_tube_rack
+  end
+
+  # Finds the tube rack purpose using the provided UUID.
+  #
+  # This method queries the TubeRack::Purpose model to find the first record
+  # that matches the given UUID. If a matching record is found, it is returned.
+  # If no matching record is found, an error is raised.
+  #
+  # @param [String] uuid The UUID of the tube rack purpose to find.
+  #
+  # @raise [StandardError] if no matching TubeRack::Purpose record is found.
+  #
+  # @return [TubeRack::Purpose] The found TubeRack::Purpose object.
+  def find_tube_rack_purpose(uuid)
+    tr_purpose = TubeRack::Purpose.with_uuid(uuid).first
+    return tr_purpose if tr_purpose
+
+    error_message = "The tube rack purpose with UUID '#{uuid}' was not found."
+    raise StandardError, error_message
+  end
+
+  # Handles the barcode assignment for a new tube rack.
+  #
+  # This method checks if a barcode already exists for the given tube rack barcode.
+  # If the barcode does not exist, it creates a new barcode for the tube rack.
+  # If the barcode already exists, it redirects the existing barcode to the new tube rack instance.
+  # This is done to allow the re-use of the physical tube rack, which has an etched barcode.
+  #
+  # @param [String] tube_rack_barcode The barcode of the tube rack.
+  # @param [TubeRack] new_tube_rack The new tube rack object to which the barcode will be assigned.
+  #
+  # @raise [StandardError] if the barcode cannot be created or redirected.
+  #
+  # @return [void]
+  def handle_tube_rack_barcode(tube_rack_barcode, new_tube_rack)
+    existing_barcode_record = Barcode.includes(:asset).find_by(barcode: tube_rack_barcode)
+
+    if existing_barcode_record.nil?
+      create_new_barcode(tube_rack_barcode, new_tube_rack)
+    else
+      redirect_existing_barcode(existing_barcode_record, new_tube_rack, tube_rack_barcode)
+    end
+  end
+
+  # Creates a new barcode for the given tube rack.
+  #
+  # This method checks if the provided tube rack barcode matches a recognized format.
+  # If the barcode format is valid, it creates a new Barcode record associated with the new tube rack.
+  # If the barcode format is not recognized, it raises an error.
+  #
+  # @param [String] tube_rack_barcode The barcode of the tube rack.
+  # @param [TubeRack] new_tube_rack The new tube rack object to which the barcode will be assigned.
+  #
+  # @raise [StandardError] if the barcode format is not recognized.
+  #
+  # @return [Barcode] The created Barcode object.
+  def create_new_barcode(tube_rack_barcode, new_tube_rack)
+    barcode_format = Barcode.matching_barcode_format(tube_rack_barcode)
+    if barcode_format.nil?
+      error_message = "The tube rack barcode '#{tube_rack_barcode}' is not a recognised format."
+      raise StandardError, error_message
+    end
+    Barcode.create!(labware: new_tube_rack, barcode: tube_rack_barcode, format: barcode_format)
+  end
```

Comment:
I thought that this kind of thing was probably replicated in the sample manifest upload code, and then found this comment - https://github.com/sanger/sequencescape/blob/a4a63a2e9b3348a3da62914739240492ed33d2e0/app/sample_manifest_excel/sample_manifest_excel/upload/processor/tube_rack.rb#L154

So looks like there might be 3 places that could do with a refactor!